### ISSUE TITLE
this.where is not a function

### ISSUE THREAD

**Issue by matthewdias** [tgriesser/knex#2619]
# Environment

Knex version: 0.14.6
Database + version: Postgres 10.3
OS: heroku-16 (Ubuntu 16.04)

# Bug

These errors are appearing and don't seem to be coming from my code.

```
TypeError: this.where is not a function
    at JoinClause.onIn (/app/node_modules/knex/lib/query/joinclause.js:116:67)
    at KnexOperation.onBuildKnex (/app/node_modules/objection/lib/queryBuilder/operations/KnexOperation.js:6:28)
    at JoinBuilder.buildInto (/app/node_modules/objection/lib/queryBuilder/QueryBuilderOperationSupport.js:237:27)
    at convertJoinBuilderFunction (/app/node_modules/objection/lib/queryBuilder/operations/ObjectionToKnexConvertingOperation.js:155:21)
    at JoinClause.convertedKnexArgumentFunction (/app/node_modules/objection/lib/queryBuilder/operations/ObjectionToKnexConvertingOperation.js:133:7)
    at Builder.join (/app/node_modules/knex/lib/query/builder.js:248:13)
    at KnexOperation.onBuildKnex (/app/node_modules/objection/lib/queryBuilder/operations/KnexOperation.js:6:28)
    at QueryBuilder.buildInto (/app/node_modules/objection/lib/queryBuilder/QueryBuilderOperationSupport.js:237:27)
    at buildInto (/app/node_modules/objection/lib/queryBuilder/QueryBuilder.js:1275:25)
    at doExecute (/app/node_modules/objection/lib/queryBuilder/QueryBuilder.js:1177:23)
    at Promise.try.then (/app/node_modules/objection/lib/queryBuilder/QueryBuilder.js:595:19)
    at tryCatcher (/app/node_modules/bluebird/js/release/util.js:16:23)
    at Promise._settlePromiseFromHandler (/app/node_modules/bluebird/js/release/promise.js:512:31)
    at Promise._settlePromise (/app/node_modules/bluebird/js/release/promise.js:569:18)
    at Promise._settlePromiseCtx (/app/node_modules/bluebird/js/release/promise.js:606:10)
    at Async._drainQueue (/app/node_modules/bluebird/js/release/async.js:138:12)
```
The only place I use `onIn()` (code works fine)
```js
...
.distinct('users.id', 'mode', 'dmChannelId')
.where('users.guildId', guildId)
.join('phrases', function() {
  this.on('users.id', 'phrases.userId')
  .onIn(raw('LOWER(phrases.text)'), phrases)
}).whereNot('phrases.userId', authorId)
...
```

**Comment by elhigu** [comment]
please add complete reproduction example. Cannot reproduce https://runkit.com/embed/hxhay295xfsu

**Comment by wubzz** [comment]
Can be reproduced when `onIn`'s second argument is an empty array.
```javascript
const Knex = require('knex');

const knex = Knex({
  client: 'pg'
});

knex('foo')
 .distinct('users.id', 'mode', 'dmChannelId')
.where('users.guildId', 1)
.join('phrases', function() {
  this.on('users.id', 'phrases.userId')
  .onIn(knex.raw('LOWER(phrases.text)'), [])
}).whereNot('phrases.userId', 4).toSQL().sql
```

**Comment by wubzz** [comment]
Fixed in #2513
